internal/worker/worker.go

```diff
@@ -27,98 +24,253 @@ type BlockResult struct {
 
-		go func(bn *big.Int) {
+	log.Debug().Msgf("Processing %d chunks of blocks", len(chunks))
+	for _, chunk := range chunks {
+		wg.Add(1)
+		go func(chunk []*big.Int) {
 			defer wg.Done()
-			result := w.processBlock(bn)
-			resultsCh <- result
-		}(blockNumber)
+			resultsCh <- w.processBatch(chunk)
+		}(chunk)
 	}
 	go func() {
 		wg.Wait()
 		close(resultsCh)
 	}()
 
-	results := make([]BlockResult, 0, blockCount)
-	for result := range resultsCh {
-		results = append(results, result)
+	results := make([]WorkerResult, 0, blockCount)
+	for batchResults := range resultsCh {
+		results = append(results, batchResults...)
 	}
 
 	return results
 }
 
-func (w *Worker) processBlock(blockNumber *big.Int) BlockResult {
-	log.Debug().Msgf("Processing block %d", blockNumber)
-	block, err := w.fetchBlock(blockNumber)
-	if err != nil {
-		return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching block %d: %v", blockNumber, err)}
+func divideIntoChunks(blockNumbers []*big.Int, batchSize int) [][]*big.Int {
+	if batchSize >= len(blockNumbers) {
+		return [][]*big.Int{blockNumbers}
 	}
+	var chunks [][]*big.Int
+	for i := 0; i < len(blockNumbers); i += batchSize {
+		end := i + batchSize
+		if end > len(blockNumbers) {
+			end = len(blockNumbers)
+		}
+		chunks = append(chunks, blockNumbers[i:end])
+	}
+	return chunks
+}
 
-	logs, err := w.fetchLogs(blockNumber)
+func (w *Worker) processBatch(blockNumbers []*big.Int) []WorkerResult {
+	blocks := w.fetchBlocks(blockNumbers)
+	logs := w.fetchLogs(blockNumbers)
+
+	var traces []TracesFetchResult
+	if w.rpc.SupportsTraceBlock {
+		traces = w.fetchTraces(blockNumbers)
+	}
+
+	return SerializeWorkerResults(blocks, logs, traces)
+}
+
+func (w *Worker) fetchBlocks(blockNumbers []*big.Int) []BlockFetchResult {
+	batch := make([]rpc.BatchElem, len(blockNumbers))
+	blockResults := make([]BlockFetchResult, len(blockNumbers))
+
+	for i, blockNum := range blockNumbers {
+		blockResults[i] = BlockFetchResult{
+			BlockNumber: blockNum,
+		}
+		batch[i] = rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{hexutil.EncodeBig(blockNum), true},
+			Result: new(map[string]interface{}),
+		}
+	}
+
+	err := w.rpc.RPCClient.BatchCallContext(context.Background(), batch)
 	if err != nil {
-		return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching logs for block %d: %v", blockNumber, err)}
+		for i := range blockResults {
+			blockResults[i].Error = err
+		}
+		return blockResults
 	}
 
-	var traces []map[string]interface{}
-	if w.rpc.SupportsTraceBlock {
-		traces, err = w.fetchTraces(blockNumber)
-		if err != nil {
-			return BlockResult{BlockNumber: blockNumber, Error: fmt.Errorf("error fetching traces for block %d: %v", blockNumber, err)}
+	for i, elem := range batch {
+		blockResults[i].BlockNumber = blockNumbers[i]
+
+		if elem.Error != nil {
+			blockResults[i].Error = elem.Error
+		} else {
+			result := *elem.Result.(*map[string]interface{})
```

Comment:
what does the `.(*map[string]interface{})` mean?

Reply:
it casts it to this type (basically an object with key value pairs)
Problem was that if I tried to use the types from go-ethereum, then batch requests did not work.
So I went for serializing them myself